# Исследование надежности заемщиков.

**Описание проекта**

Нужно разобраться, влияет ли образование, уровень заработка, семейное положение или количество детей клиента на факт погашения кредита в срок.

### Шаг 1. Обзор данных

Первым делом извлечем данные в переменную df и получем первичную информацию по данной таблице.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("A:\Practicum\Project_2\data.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [3]:
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

**По полученной иформации можно сделать следующие выводы.**
1. В колонках days_employed и total_income есть пропуски которые необходимо заполнить.
2. В колонне days_employed сумма заработка записана отрицательным числом
3. В колонне education параметры не приведены к общему слтилю.

### Шаг 2.1 Заполнение пропусков

Подсчитаем количетсво пропуском по кождой комонке

In [4]:
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Предположение подтвердилось две колонки имеют пропуски в данных.

In [5]:
df[df["total_income"].isna()]

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

Посмотрим какой процент занимают пропуски

In [6]:
df["total_income"].isna().sum() / (df["total_income"].count() + df["total_income"].isna().sum()) * 100

10.099883855981417

10% данных потеряна, необходимо их восполнить для работы с данными.

Так же надо проверить связаны ли пропуски в колннах days_employed и total_income

In [7]:
display (df[df["total_income"].isna()]["total_income"].value_counts(dropna = False))
df[df["total_income"].isna()]["days_employed"].value_counts(dropna = False)

NaN    2174
Name: total_income, dtype: int64

NaN    2174
Name: days_employed, dtype: int64

Количество пропукво в отсортированом df по колоннам равно следовательно все пропуски в данных колоннах находятся в одинаковых строках.

Вероятнее всего посчет в колонне total_income производится по формуле и одно из значений приходит из days_employed.

Следующий щаг заполнение пропусков. Поскольку значения имеют большой разброс лучше всего использовать медианное значение по колонне и им заполнить все пропуски. 

In [8]:
df["total_income"] = df["total_income"].fillna(df["total_income"].median())
display (df.isna().sum())
df.info()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Как видим колонну total_income получилось исправить.

Прежде чем исправлять колонну days_employed ее необходимо подготоваить.

### Шаг 2.2 Проверка данных на аномалии и исправления.

Что бы оценить наличие аномалий нужно посмотреть какие значени есть в колоннах children, days_employed

In [9]:
df["children"].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Данная колонна имеет два странных значения, -1 и 20. Скорее вего это опечатка и будем заменять значения -1 на 1 и 20 на 2.

In [10]:
df.loc[df["children"] == -1, ["children"]] = 1
df.loc[df["children"] == 20, ["children"]] = 2
df["children"].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

Данные исправлены, теперь можно с ними работать

In [11]:
df["days_employed"]

0         -8437.673028
1         -4024.803754
2         -5623.422610
3         -4124.747207
4        340266.072047
             ...      
21520     -4529.316663
21521    343937.404131
21522     -2113.346888
21523     -3112.481705
21524     -1984.507589
Name: days_employed, Length: 21525, dtype: float64

Колонна days_employed имеет отрицательные значения нужно это исправить.

In [12]:
df.loc[df["days_employed"] < 0, ["days_employed"]] *= (-1)
df.loc[df["days_employed"] < 0, ["days_employed"]].count()

days_employed    0
dtype: int64

Как видим теперь все значения в колонке days_employed положительные. Теперь можно заполнить пропуски медианым значением.

In [13]:
df["days_employed"] = df["days_employed"].fillna(df["days_employed"].median())
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              Среднее             1   
3          3    4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              СРЕДНЕЕ             1   
8          2    6929.865299         35               ВЫСШЕЕ             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0    2194.220567         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12  145017.937533                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

In [14]:
display (df.isna().sum())
df.info()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Как можно видеть пропусков и аномалий в колоннах days_employed и total_income больше нет, можно двигаться дальше.

### Шаг 2.3. Изменение типов данных.

Для удобства с поледущей обработкой данны не плохо перевести колону total_income в целочисленный тип

In [15]:
df["total_income"] = df["total_income"].astype(int)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  int32  
 11  purpose           21525 non-null  object 
dtypes: float64(1), int32(1), int64(5), object(5)
memory usage: 1.9+ MB


In [16]:
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

### Шаг 2.4. Удаление дубликатов.

В данных нверняка присутствуют строки-дубликаты, для их поиска необходимо все строки привести к единому стилю. 

На пример в колонне education видны разные способы написания одного и тогоже, неоюходимо исправить.

In [17]:
df["education"].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

Как видно из прердыдущей ячейки в колонне education есть что менять.

Для начал приведем все к одному регистру и оценим еще раз что нужно будет доделать.

In [18]:
df["education"] = df["education"].str.lower()
df["education"].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Теперь в строке education все впорядке. Можно дигаться дальше

Поскольку у нас нет уникальных значений для каждого клиента, дубликаты нужно искать по полному совпадению строк.
Проверим DataFrame на наличе дубликатов с помощью метода duplicated, он поможет найти одинаковые строки в df

In [19]:
df.duplicated().sum() 

71

71 строка повторяется, нам они не нужны

In [20]:
df = df.drop_duplicates().reset_index(drop=True)
df.duplicated().sum() 

0

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     21454 non-null  float64
 2   dob_years         21454 non-null  int64  
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      21454 non-null  int32  
 11  purpose           21454 non-null  object 
dtypes: float64(1), int32(1), int64(5), object(5)
memory usage: 1.9+ MB


Проблемы с колонной education могли возникнуть из-за человеческого фактора, разные люди могли по разному заносить данные в таблицу. Дубликаты же могли возникнуть на программном уровне, при переносе данных могла произойти ошибка и строки скопировались дважды.

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Для будущего удобства работы создадим два новых DF. df_education и df_family_status с кодлоннами education, education_id и family_status, family_status_id соответственно.

In [22]:
df_education = df[["education","education_id"]]
df_family_status = df[["family_status","family_status_id"]]
display(df_education.sample(10))
df_family_status.sample(10)

education  education_id
12168   среднее             1
4501    среднее             1
1762    среднее             1
14188   среднее             1
13183    высшее             0
12409   среднее             1
1885     высшее             0
15595   среднее             1
14039    высшее             0
449     среднее             1

family_status  family_status_id
8372   Не женат / не замужем                 4
7328               в разводе                 3
11754  Не женат / не замужем                 4
6213   Не женат / не замужем                 4
18941        женат / замужем                 0
7882          вдовец / вдова                 2
2847         женат / замужем                 0
7687         женат / замужем                 0
9424         женат / замужем                 0
4354         женат / замужем                 0

В таком виде от этих данных нет толку, необходимо удалить повторения в таблицах

In [23]:
df_education = df_education.drop_duplicates().reset_index(drop=True).sort_values('education_id')
df_education

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [24]:
df_family_status = df_family_status.drop_duplicates().reset_index(drop=True).sort_values('family_status_id')
df_family_status

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  Не женат / не замужем                 4

Теперь у нас есть полноценные словари со значениями и лишние столбцы из основного DF можно удалить.

In [25]:
df = df.drop(columns = ["education","family_status"])
df

children  days_employed  dob_years  education_id  family_status_id  \
0             1    8437.673028         42             0                 0   
1             1    4024.803754         36             1                 0   
2             0    5623.422610         33             1                 0   
3             3    4124.747207         32             1                 0   
4             0  340266.072047         53             1                 1   
...         ...            ...        ...           ...               ...   
21449         1    4529.316663         43             1                 1   
21450         0  343937.404131         67             1                 0   
21451         1    2113.346888         38             1                 1   
21452         3    3112.481705         38             1                 0   
21453         2    1984.507589         40             1                 0   

      gender income_type  debt  total_income                       purpose  
0          F   сотрудник     0        253875                 покупка жилья  
1          F   сотрудник     0        112080       приобретение автомобиля  
2          M   сотрудник     0        145885                 покупка жилья  
3          M   сотрудник     0        267628    дополнительное образование  
4          F   пенсионер     0        158616               сыграть свадьбу  
...      ...         ...   ...           ...                           ...  
21449      F   компаньон     0        224791             операции с жильем  
21450      F   пенсионер     0        155999          сделка с автомобилем  
21451      M   сотрудник     1         89672                  недвижимость  
21452      M   сотрудник     1        244093  на покупку своего автомобиля  
21453      F   сотрудник     0         82047         на покупку автомобиля  

[21454 rows x 10 columns]

### Шаг 2.6. Категоризация дохода.

Для дальнейшей работы с данными нужно занести каждого клиента в свою категорию по доходу. 

- 0–30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.

Необходимо создать функию которая сделает это за нас.

In [26]:
def salary_category (salary):
    if salary < 30000:
        return ("E")
    if salary < 50000:
        return ("D")
    if salary < 200000:
        return ("C")
    if salary < 1000000:
        return ("B")
    return ("A")

Функция есть, осталось ей воспользоваться, создав новый столбец в основной таблице данных.

In [27]:
df["total_income_category"] = df["total_income"].apply(salary_category)
df.head(10)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   
5         0     926.185831         27             0                 1      M   
6         0    2879.202052         43             0                 0      F   
7         0     152.779569         50             1                 0      M   
8         2    6929.865299         35             0                 1      F   
9         0    2188.756445         41             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C  
5                     B  
6                     B  
7                     C  
8                     C  
9                     C

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 11 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   children               21454 non-null  int64  
 1   days_employed          21454 non-null  float64
 2   dob_years              21454 non-null  int64  
 3   education_id           21454 non-null  int64  
 4   family_status_id       21454 non-null  int64  
 5   gender                 21454 non-null  object 
 6   income_type            21454 non-null  object 
 7   debt                   21454 non-null  int64  
 8   total_income           21454 non-null  int32  
 9   purpose                21454 non-null  object 
 10  total_income_category  21454 non-null  object 
dtypes: float64(1), int32(1), int64(5), object(4)
memory usage: 1.7+ MB


Как видим в столбце total_income_category 21454 значения, значит каждый клиент получил категорию и с этим будет проще работать.

### Шаг 2.7. Категоризация целей кредита.

Остался один шаг для основательно работы с данными. Необходимо цели кредита объеденить в категории. Посмотрим какие есть вообще цели.

In [29]:
df["purpose"].value_counts()

свадьба                                   791
на проведение свадьбы                     768
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
покупка жилья для сдачи                   651
операции с коммерческой недвижимостью     650
покупка жилья                             646
жилье                                     646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

Все эти ели можно разбить на 4 категории:
- Операции с автомобилем
- Операции с недвижимостью
- Проведение свадьбы
- Получение образования

Создадим функцию, пропустим через нее цели и добавим purpose_category столбец в таблицу.

После чего стоблец purpose можно удалить из основного df.

In [30]:
def purpose_category (purpose):
    if "свадьб" in purpose:
        return ("проведение свадьбы")
    if "автомоб" in purpose:
        return ("операции с автомобилем")
    if "образован" in purpose:
        return ("получение образования")
    return ("операции с недвижимостью")

In [31]:
df["purpose_category"] = df["purpose"].apply(purpose_category)
df = df.drop(columns = ["purpose"])
df.head(10)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   
5         0     926.185831         27             0                 1      M   
6         0    2879.202052         43             0                 0      F   
7         0     152.779569         50             1                 0      M   
8         2    6929.865299         35             0                 1      F   
9         0    2188.756445         41             1                 0      M   

  income_type  debt  total_income total_income_category  \
0   сотрудник     0        253875                     B   
1   сотрудник     0        112080                     C   
2   сотрудник     0        145885                     C   
3   сотрудник     0        267628                     B   
4   пенсионер     0        158616                     C   
5   компаньон     0        255763                     B   
6   компаньон     0        240525                     B   
7   сотрудник     0        135823                     C   
8   сотрудник     0         95856                     C   
9   сотрудник     0        144425                     C   

           purpose_category  
0  операции с недвижимостью  
1    операции с автомобилем  
2  операции с недвижимостью  
3     получение образования  
4        проведение свадьбы  
5  операции с недвижимостью  
6  операции с недвижимостью  
7     получение образования  
8        проведение свадьбы  
9  операции с недвижимостью

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 11 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   children               21454 non-null  int64  
 1   days_employed          21454 non-null  float64
 2   dob_years              21454 non-null  int64  
 3   education_id           21454 non-null  int64  
 4   family_status_id       21454 non-null  int64  
 5   gender                 21454 non-null  object 
 6   income_type            21454 non-null  object 
 7   debt                   21454 non-null  int64  
 8   total_income           21454 non-null  int32  
 9   total_income_category  21454 non-null  object 
 10  purpose_category       21454 non-null  object 
dtypes: float64(1), int32(1), int64(5), object(4)
memory usage: 1.7+ MB


Теперь DataFrame полностью готов для работы с ним. Пропусков нет, дубликаты удалены, лищние таблицы убраны в отделбную базу, клиенты собраны в свои категории по зарплате и целям на кредит.

### Ответы на вопросы.

К данной таблице были поставлены следущие вопросы:
1. Есть ли зависимость между количеством детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

##### Вопрос 1:

Для ответа на данный вопрос необходимо понимать какие значения принимают данные.

In [33]:
df["debt"].value_counts()


0    19713
1     1741
Name: debt, dtype: int64

In [34]:
df["children"].value_counts()

0    14091
1     4855
2     2128
3      330
4       41
5        9
Name: children, dtype: int64

Создадим сводную таблицу которая будет принимать необходимые значения.

In [35]:
children_debt = df.pivot_table(index = ["children"], columns=["debt"], values = "gender",  aggfunc="count")
children_debt

debt            0       1
children                 
0         13028.0  1063.0
1          4410.0   445.0
2          1926.0   202.0
3           303.0    27.0
4            37.0     4.0
5             9.0     NaN

Заменим NaN на 0 поскольку семьи с 5 детьми не совершали просрочек.

In [36]:
children_debt.loc[5,1] = 0
children_debt

debt            0       1
children                 
0         13028.0  1063.0
1          4410.0   445.0
2          1926.0   202.0
3           303.0    27.0
4            37.0     4.0
5             9.0     0.0

Для оценки будем использовать процентное отношение 

In [37]:
children_debt["prosetn_of_debt"] = children_debt[1] / (children_debt[1] + children_debt[0]) * 100
children_debt

debt            0       1  prosetn_of_debt
children                                  
0         13028.0  1063.0         7.543822
1          4410.0   445.0         9.165808
2          1926.0   202.0         9.492481
3           303.0    27.0         8.181818
4            37.0     4.0         9.756098
5             9.0     0.0         0.000000

In [38]:
children_debt["prosetn_of_debt"].max() - children_debt["prosetn_of_debt"][:5].min()

2.2122752630549503

##### Вывод 1:

Можно сделать вывод что просрочка не зависит от количества детей, разница между минимальным и максимальным значением состовляет всего 2.2% (в расчет не берем семьи с 5 детьми, слишком малло данных)

##### Вопрос 2:

Здесь будем действовать аналогично как и в предыдущем вопросе.

In [39]:
family_status_id_debt = df.pivot_table(index = ["family_status_id"], columns=["debt"], values = "gender",  aggfunc="count")
family_status_id_debt

debt                  0    1
family_status_id            
0                 11408  931
1                  3763  388
2                   896   63
3                  1110   85
4                  2536  274

In [40]:
family_status_id_debt["prosetn_of_debt"] = family_status_id_debt[1] / (family_status_id_debt[1] + family_status_id_debt[0]) * 100
family_status_id_debt

debt                  0    1  prosetn_of_debt
family_status_id                             
0                 11408  931         7.545182
1                  3763  388         9.347145
2                   896   63         6.569343
3                  1110   85         7.112971
4                  2536  274         9.750890

In [41]:
family_status_id_debt["prosetn_of_debt"].max() - family_status_id_debt["prosetn_of_debt"].min()

3.1815466140218707

##### Вывод 2

Можно сделать вывод что просрочка не зависит от семейного положения, разница между минимальным и максимальным значением состовляет всего 3.2%

##### Вопрос 3

Отвеаем на вопрос все по томуже принципу.

In [42]:
total_income_category_debt = df.pivot_table(index = ["total_income_category"], columns=["debt"], values = "gender",  aggfunc="count")
total_income_category_debt

debt                       0     1
total_income_category             
A                         23     2
B                       4685   356
C                      14656  1360
D                        329    21
E                         20     2

In [43]:
total_income_category_debt["prosetn_of_debt"] = total_income_category_debt[1] / (total_income_category_debt[1] + total_income_category_debt[0]) * 100
total_income_category_debt

debt                       0     1  prosetn_of_debt
total_income_category                              
A                         23     2         8.000000
B                       4685   356         7.062091
C                      14656  1360         8.491508
D                        329    21         6.000000
E                         20     2         9.090909

In [44]:
total_income_category_debt["prosetn_of_debt"].max() - total_income_category_debt["prosetn_of_debt"].min()

3.0909090909090917

##### Вывод 3

Можно сделать вывод что просрочка не зависит от заработка, разница между минимальным и максимальным значением состовляет всего 3.1%

##### Вопрос 4

Действуем аналогично предыдущим вопросам.

In [45]:
purpose_category_debt = df.pivot_table(index = ["purpose_category"], columns=["debt"], values = "gender",  aggfunc="count")
purpose_category_debt

debt                          0    1
purpose_category                    
операции с автомобилем     3903  403
операции с недвижимостью  10029  782
получение образования      3643  370
проведение свадьбы         2138  186

In [46]:
purpose_category_debt["prosetn_of_debt"] = purpose_category_debt[1] / (purpose_category_debt[1] + purpose_category_debt[0]) * 100
purpose_category_debt

debt                          0    1  prosetn_of_debt
purpose_category                                     
операции с автомобилем     3903  403         9.359034
операции с недвижимостью  10029  782         7.233373
получение образования      3643  370         9.220035
проведение свадьбы         2138  186         8.003442

In [47]:
purpose_category_debt["prosetn_of_debt"].max() - purpose_category_debt["prosetn_of_debt"].min()

2.125660490212206

##### Вывод 4

Можно сделать вывод что просрочка не зависит от цели кредита, разница между минимальным и максимальным значением состовляет всего 2.1%

## Общий вывод:

Проверив 4 гипотизы получили что возвращение кредитных средств в срок не зависят от семейного положения, количесвта детей, цели кредита и заработка. 

Так же можно предположить что данных не достаточно для определенных выводов, по скольку в некоторых категорияю количество людей в районе 1000, этого не достаточно для однозначных выводов.

Можно предположить, что самым не надежным плательщиком является человек вдовец/вдова без детей, с заработком 30-50т.р. и цель операции с недвижимостью.